core/src/apps/cardano/layout.py

```diff
@@ -116,18 +131,80 @@ async def show_warning_tx_staking_key_hash(
     await require_confirm(ctx, Paginated([t1, t2, t3]))
 
 
-async def confirm_transaction(ctx, amount: int, fee: int, protocol_magic: int):
+async def confirm_transaction(
+    ctx,
+    amount: int,
+    fee: int,
+    protocol_magic: int,
+    certificates: List[CardanoTxCertificateType],
+    withdrawals: List[CardanoTxWithdrawalType],
+):
+    pages = []
+
     t1 = Text("Confirm transaction", ui.ICON_SEND, ui.GREEN)
     t1.normal("Transaction amount:")
     t1.bold(format_coin_amount(amount))
     t1.normal("Transaction fee:")
     t1.bold(format_coin_amount(fee))
+    pages.append(t1)
 
-    t2 = Text("Confirm transaction", ui.ICON_SEND, ui.GREEN)
-    t2.normal("Network:")
-    t2.bold(protocol_magics.to_ui_string(protocol_magic))
+    if len(certificates) > 0:
+        t2 = Text("Confirm transaction", ui.ICON_SEND, ui.GREEN)
+        t2.normal("Certificates:")
+        for certificate in certificates:
+            t2.bold(CERTIFICATE_TYPE_NAMES[certificate.type])
+        pages.append(t2)
+
+    t3 = Text("Confirm transaction", ui.ICON_SEND, ui.GREEN)
+    if len(withdrawals) > 0:
+        t3.bold("%s withdrawals" % len(withdrawals))
+    t3.normal("Network:")
+    t3.bold(protocol_magics.to_ui_string(protocol_magic))
+    pages.append(t3)
+
+    await require_hold_to_confirm(ctx, Paginated(pages))
```

Comment:
I am not sure what is the Trezor preferred way but I would prefer separating layour generation from side-effect

Reply:
I am not sure how else I would solve this, but see the comment below - maybe we can remove it altogether.